### ISSUE TITLE
FieldArray: fields.remove(i) does not completely remove field

### ISSUE THREAD

**Issue by dirkroorda** [erikras/redux-form#2841]
### What is the current behavior?

I'm using a FieldArray, copied from the FieldArray example (the "hobbies").
I can remove fields by pressing the button that triggers the fields.remove(i) action.
That works fine, and I see correct values in the form state.

But when I submit the form, (I have 
```  
  destroyOnUnmount: false,
  enableReinitialize: true,
  keepDirtyOnReinitialize: true,
```
the next screen shows the removed field with empty contents.

### What is the expected behavior?
After a submit, the removed field should not show up.
By the way, when I set `keepDirtyOnReinitialize: false`, this problem does not occur.
However, I need `keepDirtyOnReinitialize: true`, because I let the user unmount and remount forms while editing.

### Observations
When I look into the form state, I see in the values slice a null in the place of the removed array item.
The initial values are correct.

During the submit process the correct values are handed over to the database.
After the submit, the re-initialize brings the removed field to life, with null contents.

Looking back in the state, I see
```@@redux-form/SET_SUBMIT_SUCCEEDED```
resulting in a correct state (I'm looking at the `values` slice).
Then there is a bunch of (UN)REGISTER_FIELD actions, after which the values are still correct.

Then an action of mine reports a successful update in the database, and triggers a
```@@redux-form/INITIALIZE```
with the payload still having the correct values.
But the resulting state after this action has the extra null in the values.

So this is the first action where it happens.
Maybe the cause is that the extra field lingers on the the `fields` and `registeredFields` slices, and the reducer reintroduces them into values upon re-initialization.

### What's your environment?
redux-form 6.6.3, running on Chrome 57.

**Comment by dirkroorda** [comment]
I found a workaround!
After a successful submit, issue an extra reset.
Basically, this prevents the removed field from registering again.

There will be an extra update because of this, and I hope the root cause will be solved some time.

This is the way I did it. `ItemForm` is my inner form component, and I pass it an `onSubmitSuccess` callback when I wrap `ItemForm` with the `reduxForm()` call:

```
const onSubmitSuccess = (result, dispatch, { reset }) => {
  reset()
}

export default connect(getTables, { mod: modItem })(reduxForm({
  destroyOnUnmount: false,
  enableReinitialize: true,
  keepDirtyOnReinitialize: true,
  onSubmitSuccess,
})(ItemForm))
```

The `getTables, { mod: modItem }` part is not relevant to this problem. They help `ItemForm` to connect with the part of the state that contains the data fetched from the database. `modItem` sends updated items to the database.
